// png-async/lib/index.js
/// <reference path="../typings/index.d.ts" />
'use strict';
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var stream = require('stream');
var Parser = require('./parser');
var Packer = require('./packer');
(function (EDeflateStrategy) {
    EDeflateStrategy[EDeflateStrategy["DEFAULT_STRATEGY"] = 0] = "DEFAULT_STRATEGY";
    EDeflateStrategy[EDeflateStrategy["FILTERED"] = 1] = "FILTERED";
    EDeflateStrategy[EDeflateStrategy["HUFFMAN_ONLY"] = 2] = "HUFFMAN_ONLY";
    EDeflateStrategy[EDeflateStrategy["RLE"] = 3] = "RLE";
    EDeflateStrategy[EDeflateStrategy["FIXED"] = 4] = "FIXED";
})(exports.EDeflateStrategy || (exports.EDeflateStrategy = {}));
var EDeflateStrategy = exports.EDeflateStrategy;
(function (EFilterType) {
    EFilterType[EFilterType["Auto"] = -1] = "Auto";
    EFilterType[EFilterType["None"] = 0] = "None";
    EFilterType[EFilterType["Sub"] = 1] = "Sub";
    EFilterType[EFilterType["Up"] = 2] = "Up";
    EFilterType[EFilterType["Average"] = 3] = "Average";
    EFilterType[EFilterType["Paeth"] = 4] = "Paeth";
})(exports.EFilterType || (exports.EFilterType = {}));
var EFilterType = exports.EFilterType;
function createImage(option) {
    return new Image(option);
}
exports.createImage = createImage;
var Image = (function (_super) {
    __extends(Image, _super);
    function Image(option) {
        var _this = this;
        if (option === void 0) { option = {}; }
        _super.call(this);
        this.width = option.width || 0;
        this.height = option.height || 0;
        if (this.width > 0 && this.height > 0) {
            this.data = new Buffer(4 * this.width * this.height);
            if (option.fill) {
                this.data.fill(0);
            }
        }
        else {
            this.data = null;
        }
        this.gamma = 0;
        this.writable = true;
        this._parser = new Parser(option || {});
        this._parser.on('error', this.emit.bind(this, 'error'));
        this._parser.on('close', this._handleClose.bind(this));
        this._parser.on('metadata', this._metadata.bind(this));
        this._parser.on('gamma', this._gamma.bind(this));
        this._parser.on('parsed', function (data) {
            _this.data = data;
            _this.emit('parsed', data);
        });
        this._packer = new Packer(option);
        this._packer.on('data', this.emit.bind(this, 'data'));
        this._packer.on('end', this.emit.bind(this, 'end'));
        this._packer.on('close', this._handleClose.bind(this));
        this._packer.on('error', this.emit.bind(this, 'error'));
    }
    Image.prototype.pack = function () {
        var _this = this;
        setImmediate(function () {
            _this._packer.pack(_this.data, _this.width, _this.height);
            _this.readable = true;
        });
        return this;
    };
    Image.prototype.parse = function (data, callback) {
        var _this = this;
        if (callback) {
            var onParsed_1 = null, onError_1 = null;
            this.once('parsed', onParsed_1 = function (data) {
                _this.removeListener('error', onError_1);
                _this.data = data;
                callback(null, _this);
            });
            this.once('error', onError_1 = function (err) {
                _this.removeListener('parsed', onParsed_1);
                callback(err, null);
            });
        }
        this.end(data);
        return this;
    };
    Image.prototype._write = function (data, encoding, callback) {
        return this._parser._write(data, encoding, callback);
    };
    Image.prototype.end = function (data) {
        return this._parser.end(data);
    };
    Image.prototype.bitblt = function (dst, sx, sy, w, h, dx, dy) {
        if (sx > this.width || sy > this.height || sx + w > this.width || sy + h > this.height) {
            throw new Error('bitblt reading outside image');
        }
        if (dx > dst.width || dy > dst.height || dx + w > dst.width || dy + h > dst.height) {
            throw new Error('bitblt writing outside image');
        }
        for (var y = 0; y < h; y++) {
            this.data.copy(dst.data, ((dy + y) * dst.width + dx) << 2, ((sy + y) * this.width + sx) << 2, ((sy + y) * this.width + sx + w) << 2);
        }
        return this;
    };
    Image.prototype._read = function () {
    };
    Image.prototype._metadata = function (metadata) {
        this.width = metadata.width;
        this.height = metadata.height;
        this.data = metadata.data;
        delete metadata.data;
        this.emit('metadata', metadata);
    };
    Image.prototype._gamma = function (gamma) {
        this.gamma = gamma;
    };
    Image.prototype._handleClose = function () {
        if (!this._parser.writable && !this._packer.readable) {
            this.emit('close');
        }
    };
    return Image;
}(stream.Duplex));
exports.Image = Image;
